Storage/SQL/SQLiteRemoteClientsAndTabs.swift

```diff
@@ -249,77 +233,51 @@ open class SQLiteRemoteClientsAndTabs: RemoteClientsAndTabs {
-        }
+        return db.withConnection { conn -> [ClientAndTabs] in
+            let clientsCursor = conn.executeQuery("SELECT * FROM \(TableClients) WHERE EXISTS (SELECT 1 FROM \(TableRemoteDevices) rd WHERE rd.guid = fxaDeviceId) ORDER BY modified DESC", factory: SQLiteRemoteClientsAndTabs.remoteClientFactory)
+            let tabsCursor = conn.executeQuery("SELECT * FROM \(TableTabs) WHERE client_guid IS NOT NULL ORDER BY client_guid DESC, last_used DESC", factory: SQLiteRemoteClientsAndTabs.remoteTabFactory)
 
-        log.debug("Found \(clients.count) clients in the DB.")
+            defer {
+                clientsCursor.close()
+                tabsCursor.close()
+            }
 
-        var acc = [String: [RemoteTab]]()
-        let deferredTabsCursor = db.withConnection { connection -> Cursor<RemoteTab> in
-            connection.executeQuery("SELECT * FROM \(TableTabs) WHERE client_guid IS NOT NULL ORDER BY client_guid DESC, last_used DESC", factory: SQLiteRemoteClientsAndTabs.remoteTabFactory)
-        }
-        
-        guard let tabCursor = deferredTabsCursor.value.successValue else {
-            return deferMaybe(DatabaseError(err: deferredTabsCursor.value.failureValue as NSError?))
-        }
-        
-        defer {
-            tabCursor.close()
-        }
-        
-        // Aggregate clientGUID -> RemoteTab.
-        for tab in tabCursor {
-            if let tab = tab, let guid = tab.clientGUID {
-                if acc[guid] == nil {
-                    acc[guid] = [tab]
+            var acc = [String: [RemoteTab]]()
+            for tab in tabsCursor {
+                if let tab = tab, let guid = tab.clientGUID {
+                    if acc[guid] == nil {
+                        acc[guid] = [tab]
+                    } else {
+                        acc[guid]!.append(tab)
+                    }
                 } else {
-                    acc[guid]!.append(tab)
+                    log.error("Couldn't cast tab (\(tab ??? "nil")) to RemoteTab.")
                 }
-            } else {
-                log.error("Couldn't cast tab (\(tab ??? "nil")) to RemoteTab.")
             }
-        }
 
-        let deferred = Deferred<Maybe<[ClientAndTabs]>>(defaultQueue: DispatchQueue.main)
+            // Most recent first.
+            let fillTabs: (RemoteClient) -> ClientAndTabs = { client in
+                var tabs: [RemoteTab]? = nil
+                if let guid: String = client.guid {
+                    tabs = acc[guid]
+                }
+                return ClientAndTabs(client: client, tabs: tabs ?? [])
+            }
 
-        // Most recent first.
-        let fillTabs: (RemoteClient) -> ClientAndTabs = { client in
-            var tabs: [RemoteTab]? = nil
-            if let guid: String = client.guid {
-                tabs = acc[guid]
+            let removeLocalClient: (RemoteClient) -> Bool = { client in
```

Comment:
We shouldn't need this. Not only because if we know the local client is the one with a null GUID, we should just filter it out in the SQL, but also this should never be the case — `guid` is `guid TEXT PRIMARY KEY`!

Step through this with the debugger. It should never do anything useful.